### DISCUSSION TITLE
Refactoring of the menu code

### DISCUSSION THREAD

**Discussion by agalakhov** [Jyers/Marlin#828] [Ideas]
JFYI, I'm currently trying to greatly simplify the current menu code. I'm focusing on the rewriting `CrealityDWINClass::Menu_Item_Handler()` function. The expected result is:
* the new code is expected to be 3x shorter than the existing one
* menu items with simple actions will require single line of code
* returning from submenus will keep selection _and_ scrolling position
* it will be possible to make some submenus with icons like the main menu
* I expect the new code to use less flash space

I announce this in order not to interfere with someone's else work. Let's discuss alternative architecture proposals, if any. I hope to make some early architecture preview available soon.

**Comment by Jyers** [comment]
Sounds great! From what you've already done you seem to have much more C/C++ experience that I do. Can't wait to see your implementation. Feel free to contact me if you want to discuss or anything of the like.

**Reply by agalakhov** [reply]
Thank you! I believe I quite understand how your code works. My current goal is to write the same as simple as possible.

You can track my progress at https://github.com/agalakhov/Marlin/tree/menu-refactor

I'm in C++ since 1998 :) But now I'm mostly moving to Rust since C++ has too many strange things. It's too easy to shoot your own foot like I did yesterday.

**Reply by Jyers** [reply]
You should have seen the old dwin code 😅 https://github.com/MarlinFirmware/Marlin/blob/2.0.x/Marlin/src/lcd/dwin/e3v2/dwin.cpp

I've been using C++ since 2012 (albeit I was 10), so you've definitely got a few years on me, both in programming and life experience. I will definitely be following your progress. If you ever want to chat, be it email or call or whatnot, just let me know. I can only imagine what I could learn from you and it would be good to make sure I fully understand your methodology if I am to continue to maintain the code.

**Reply by agalakhov** [reply]
Don't hesitate to ask any C++ questions if you have any!

I'm currently splitting the big class in smaller ones. A class basically costs nothing, so this helps to organize the code. I'll have (at least) a menu class which tracks menu state and renders menus, a geometry class which knows all coordinates (and this will be the only place to change for the landscape mode) and a settings class for fonts, colors and other EEPROM stuff. The current `CrealityDWINClass` will contain only things like leveling and others that do not belong to the above ones.

My goal is to make the code _declarative_. That is, instead of writing function calls we will populate constant arrays. The menu code, for example, will look like
```
const Menu control_menu = {
  "Control",
  {
    { Icon::Temperature, "Temperature", SubMenu(temperature_menu) },
    { Icon::Motion, "Motion", SubMenu(motion_menu) },
    // ...
    { Icon::Save, "Store Settings", Function(Save_Settings) },
    EndMenu
  }
};
```

**Reply by Jyers** [reply]
@agalakhov By the way, is there any way I can help with your effort? I'm don't have any huge feature plan's at the moment (apart from landscape orientation for which I'm going to wait for the refactoring to be done) so I would absolutely love to lend a hand if you think I could be useful. Feel free to contact me at jyers.marlin@gmail.com if there's any way I can help!

**Reply by agalakhov** [reply]
Right now nobody can help, sorry :) After some time testing will be necessary. And you can already start to figure out correct coordinates for the landscape mode. I already moved most of coordinates magic to separate file.

**Reply by Jyers** [reply]
Alright, well if you do end up needing some grunt coding work done (for the menu items for instance) I would happy to help if you think tag teaming it might speed up the process. Either way great work, can't wait to see how it comes out.

**Comment by ghost** [comment]
I'm trying to test the code given by @tititopher68 on my feature request  #815. Can this code work with your implementation or does it need to be rewritten?

**Reply by Jyers** [reply]
I can tell you right now that it will most likely need to be rewritten, but you'll be fine for a while until this gets implemented

**Reply by ghost** [reply]
Okay, Thanks.

**Reply by ghost** [reply]
I understand that. I don't mean to imply that it was finished or anything like that, I was merely trying to avoid diving head first into something i have very little experience in, after which I would find out I can't use any of it.

By the way, English isn't my first language, so forgive me if I'm not always easy to understand.

**Comment by agalakhov** [comment]
Ok, while still not functional (it compiles and even runs but uses mostly old code), the new menu structure is already visible. This is how I describe menus now:
https://github.com/agalakhov/Marlin/blob/menu-refactor/Marlin/src/lcd/dwin/creality_menu.cpp

And this is the screen geometry:
https://github.com/agalakhov/Marlin/blob/menu-refactor/Marlin/src/lcd/dwin/geometry.h#L71

**Reply by dwery** [reply]
Looks very clean, keep up the good work!

**Comment by agalakhov** [comment]
Sorry for the delay, now I'm back to work. @Jyers , what are your plans for further development?

**Reply by dwery** [reply]
Most of @Jyers code has been upstreamed. Your refactoring could easily be too, as it allows very tidy code.
